### PR TITLE
Fix leak when early returning out of `box` syntax

### LINKED ISSUE #62289: Leak when returning out of box syntax
@nils-ohlmeier mentioned a leak coming from `vec![f()?]` but not `vec![f()?; 1]` [here](https://github.com/mozilla/webrtc-sdp/blob/master/src/attribute_type.rs#L2158-L2160), I think I tracked this down to the use of `box` in `vec!` when given a comma-separated list.

I've made [a smallish test](https://gist.github.com/agashlin/cd5e2227f13d48e3c701b68df4f89e6c) that demonstrates it with `stats_alloc`, if it gets much simpler than this then the leak doesn't occur when built with --release.

I'm testing on rustc 1.37.0-nightly (0af8e872e 2019-06-30) nightly-x86_64-pc-windows-msvc, targeting x86_64-pc-windows-msvc.

### PR DESCRIPTION
**Pull request by wesleywiser**
Fixes #62289

r? @matthewjasper

### CODE REVIEW

File: src/test/mir-opt/issue-62289.rs [hunk 1]
```diff
@@ -0,0 +1,82 @@
+// check that we clear the "ADT master drop flag" even when there are
+// no fields to be dropped.
```
Comment (matthewjasper) [inline src/test/mir-opt/issue-62289.rs]: This comment appears to have been copied from another test.

File: src/test/mir-opt/issue-62289.rs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+fn main() {
+    test();
+}
+
+// END RUST SOURCE
+// START rustc.test.ElaborateDrops.before.mir
+// fn test() -> std::option::Option<std::boxed::Box<u32>> {
+//     ...
+//     bb0: {
+//         ...
```
Comment (matthewjasper) [inline src/test/mir-opt/issue-62289.rs]: Is there a reason for omitting lines in so many places?

Comment (wesleywiser) [reply src/test/mir-opt/issue-62289.rs]: I left out all the `StorageLive(...)` and `StorageDead(...)` lines but I can add them back in if you'd prefer a more exact test.

Comment (matthewjasper) [reply src/test/mir-opt/issue-62289.rs]: I don't think there's any reason to omit them, given that the test is probably very sensitive to MIR generation changes without them.

**Comment by wesleywiser** [comment]
@matthewjasper Fixed!

**Comment by Centril** [comment]
Failed in https://github.com/rust-lang/rust/pull/62399, @bors r- rollup=never

**Comment by wesleywiser** [comment]
Rebased; all tests pass locally.

@bors r=matthewjasper

**Comment by matthewjasper** [comment]
@bors r- failure was legit due to wasm using `panic=abort`.
You'll need to either add `// ignore wasm32-bare` or use an earlier pass such as `SimplifyCfg-qualify-consts.after`.

**Comment by wesleywiser** [comment]
Gotcha. I guess I don't have a wasm environment setup locally? Are there docs anywhere I can read on setting that up? I don't see anything in the rustc guide.

**Comment by wesleywiser** [comment]
Added `// ignore wasm32-bare`

File: src/test/mir-opt/issue-62289.rs [hunk 3]
```diff
@@ -0,0 +1,91 @@
+// check that we don't forget to drop the Box if we early return before
+// initializing it
+// ignore-tidy-linelength
+// ignore wasm32-bare
```
Comment (matthewjasper) [inline src/test/mir-opt/issue-62289.rs]:
```suggestion
// ignore-wasm32-bare
```

**Pull request merged**